mesonbuild/mintro.py

```diff
@@ -125,7 +130,169 @@ def list_target_files(target_name, coredata, builddata):
+            else:
+                (comp, _) = self.detect_compilers(lang, need_cross_compiler)
+            if comp is None:
+                return False
+            self.add_base_options(comp)
+        return True
+
+    def func_project(self, node, args, kwargs):
+        if len(args) < 1:
+            raise InvalidArguments('Not enough arguments to project(). Needs at least the project name.')
+
+        proj_langs = self.flatten_args(args[1:])
+
+        if os.path.exists(self.option_file):
+            oi = optinterpreter.OptionInterpreter(self.subproject)
+            oi.process(self.option_file)
+            self.coredata.merge_user_options(oi.options)
+
+        def_opts = kwargs.get('default_options', [])
+        if isinstance(def_opts, mparser.ArrayNode):
+            def_opts = [x.value for x in def_opts.args.arguments]
+
+        self.project_default_options = mesonlib.stringlistify(def_opts)
+        self.project_default_options = cdata.create_options_dict(self.project_default_options)
+        self.default_options.update(self.project_default_options)
+
+        if not self.is_subproject() and 'subproject_dir' in kwargs:
+            spdirname = kwargs['subproject_dir']
+            if not isinstance(spdirname, str):
+                raise RuntimeError('Subproject_dir must be a string')
+            if os.path.isabs(spdirname):
+                raise RuntimeError('Subproject_dir must not be an absolute path.')
+            if spdirname.startswith('.'):
+                raise RuntimeError('Subproject_dir must not begin with a period.')
+            if '..' in spdirname:
+                raise RuntimeError('Subproject_dir must not contain a ".." segment.')
+            self.subproject_dir = spdirname
+        if not self.is_subproject():
+            subprojects_dir = os.path.join(self.source_root, self.subproject_dir)
+            if os.path.isdir(subprojects_dir):
+                for i in os.listdir(subprojects_dir):
+                    if os.path.isdir(os.path.join(subprojects_dir, i)):
+                        self.do_subproject(i)
+
+        self.coredata.init_backend_options(self.backend.name)
+        options = {k: v for k, v in self.environment.cmd_line_options.items() if k.startswith('backend_')}
+
+        self.coredata.set_options(options)
+        self.add_languages(proj_langs)
+
+    def do_subproject(self, dirname):
+        if dirname in self.subproject_stack:
+            fullstack = self.subproject_stack + [dirname]
+            incpath = ' => '.join(fullstack)
+            raise RuntimeError('Recursive include of subprojects: %s.' % incpath)
+        subproject_dir_abs = os.path.join(self.environment.get_source_dir(), self.subproject_dir)
+        subpr = os.path.join(subproject_dir_abs, dirname)
+        self.global_args_frozen = True
+        try:
+            subi = BuildoptionsInterperter(subpr, '', self.backend.name, cross_file=self.cross_file, subproject=dirname, subproject_dir=self.subproject_dir)
```

Comment:
However, this would not cover the case where a new language is added in a subproject. The options for this language would be missing in this case. So I think that keeping this extra logic is worth it.

Reply:
indeed. I didn't think of that. The language detection is useful.
And i think it's not a big problem that the IDE will not know if that language is used by the main project or a subproject.

So i agree we should keep this.